Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,396 @@
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var branch = repo.Branches[branchName];
+                if (branch == null) return false;
+                return repo.Refs.ReachableFrom(
+                              new [] {repo.Refs[branch.CanonicalName]},
+                              new [] {repo.Lookup<Commit>(commitOrBranchName)}).Any();
+            }
+        }
+
+        public IEnumerable<string> ListFiles(string path, SearchOption searchOption, params string[] lookupList)
+        {
+            path = PathUtility.CleanPath(path);
+
+            if (!FileSystemHelpers.IsSubfolder(RepositoryPath, path))
+            {
+                throw new NotSupportedException("Only paths relative to the repository root path are supported, path provided: '{0}' is not a child folder of '{1}'".FormatCurrentCulture(path, RepositoryPath));
+            }
+
+            if (!Directory.Exists(path)) return Enumerable.Empty<string>();
+
+            var lookupPatterns = lookupList.Select(l => string.Format("^{0}$", Regex.Escape(l).Replace("\\*", ".*")));
+
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var files = repo.Diff.Compare<TreeChanges>(null, DiffTargets.Index, lookupList, compareOptions: new CompareOptions(){IncludeUnmodified = true, Similarity = SimilarityOptions.None})
+                                      .Select(d => Path.Combine(repo.Info.WorkingDirectory, d.Path))
+                                      .Where(p => p.StartsWith(path, StringComparison.OrdinalIgnoreCase));
+
+                switch (searchOption)
+                {
+                    case SearchOption.TopDirectoryOnly:
+                        files = files.Where(line => !line.Substring(path.Length).TrimStart('\\').Contains('\\'));
+                        break;
+
+                    case SearchOption.AllDirectories:
+                        break;
+
+                    default:
+                        throw new NotSupportedException("Search option {0} is not supported".FormatCurrentCulture(searchOption));
+                }
+                // Make sure to materialize the list before finalizing repo
+                return files.ToList();
+            }
+        }
+
+        private string PostReceiveHookPath
+        {
+            get
+            {
+                return Path.Combine(RepositoryPath, ".git", "hooks", "post-receive");
+            }
+        }
+
+        public bool SkipPostReceiveHookCheck
+        {
+            get;
+            set;
+        }
+
+        public bool Exists
+        {
+            get
+            {
+                if(!SkipPostReceiveHookCheck && !File.Exists(PostReceiveHookPath))
+                {
+                    return false;
+                }
+
+                if (LibGit2Sharp.Repository.IsValid(RepositoryPath) &&
+                    !RepositoryPath.TrimEnd(new [] {'/', '\\'}).EndsWith(".git"))
```

Comment:
The old function in `GitExeRepository` will only return true if `RepositoryPath` is the path to the root of the repo working directory, not the path to the `.git` folder or any sub-path in the repo.
`Repository.IsValid()` returns true for root of working directory or the .git folder, hence the second line that makes sure it's not the .git folder.

That's said, the old function will actually throw if the repo happens to be corrupted. Have we faced that issue a lot in the past? LibGit2Sharp doesn't seem to throw if the repo is corrupted since we are not doing any operations on the repo.